src/main.zig

```diff
@@ -393,19 +486,66 @@ pub fn Deserializer(comptime ReaderType: type) type {
                 else => return error.MismatchingFormatType,
             };
 
-            data_type.* = try reader.readIntBig(i8);
+            var _data_type = try reader.readIntBig(i8);
+
             const buffer = try self.allocator.alloc(u8, len);
-            const actual_len = try reader.readAll(buffer);
+            errdefer self.allocator.free(buffer);
 
+            const actual_len = try reader.readAll(buffer);
             if (actual_len < len) return error.EndOfStream;
 
+            // Do not replace the value of input variable `data_type` unless we
+            // are certain this function call will not fail
+            data_type.* = _data_type;
+
             return buffer;
         }
 
-        /// Resets the internal buffer of `Self`. Calling any of the deserializing functions
-        /// after this will rewrite the buffer starting at index 0.
-        pub fn reset(self: *Self) void {
-            self.index = 0;
+        /// Recursively frees any value that was returned by a `deserialize` function (or any of it's specific variants)
+        pub fn free(self: *Self, ptr: anytype) void {
+            const T = @TypeOf(ptr);
+            comptime assert(trait.is(.Pointer)(T));
+
+            const C = comptime meta.Child(T);
+
+            if (comptime trait.hasFn("deinit")(C)) return C.deinit(ptr, self);
+
+            switch (C) {
+                []const u8, []u8 => {
+                    self.allocator.free(ptr.*);
+                },
+                else => switch (@typeInfo(C)) {
+                    .Struct => {
+                        inline for (meta.fields(C)) |struct_field| {
+                            self.free(&@field(ptr, struct_field.name));
+                        }
+                    },
+                    .Array => {
+                        for (ptr.*) |*v| {
+                            self.free(v);
+                            // Since this is a static array, we do not need to free
+                            // the array itself, only it's individual elements
+                        }
+                    },
+                    .Pointer => {
+                        return switch (@typeInfo(C).Pointer.size) {
+                            .One => {
+                                self.free(ptr.*);
+                                self.allocator.destroy(ptr.*);
+                            },
+                            .Slice, .Many => {
+                                for (ptr.*) |*v| {
+                                    self.free(v);
+                                }
+                                self.allocator.free(ptr.*);
```

Comment:
A many-pointer (`[*]T`) is not an acceptable argument to be passed to `free`. The allocator must know the size to successfully free all the memory. In my opinion, we shouldn't accept many-pointers at all for both serializing and deserializing as the msg_pack format always requires a length.

Reply:
You are right. I will change the code to treat Many pointers the same as C pointers, and throw a compiler error instead. :)